File: trees/BinaryTree.java
// File Name: BinaryTree.java
import java.util.*;

class Node {
    int data;
    Node left, right;

    Node(int data) {
        this.data = data;
        left = right = null;
    }
}

public class BinaryTree {
    Node root;

    void insert(int data) {
        Node newNode = new Node(data);
        if (root == null) {
            root = newNode;
            return;
        }

        Queue<Node> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty()) {
            Node temp = q.poll();

            if (temp.left == null) {
                temp.left = newNode;
                return;
            } else q.add(temp.left);

            if (temp.right == null) {
                temp.right = newNode;
                return;
            } else q.add(temp.right);
        }
    }

    boolean search(Node root, int key) {
        if (root == null) return false;
        if (root.data == key) return true;
        return search(root.left, key) || search(root.right, key);
    }

    void inorder(Node root) {
        if (root != null) {
            inorder(root.left);
            System.out.print(root.data + " ");
            inorder(root.right);
        }
    }

    void preorder(Node root) {
        if (root != null) {
            System.out.print(root.data + " ");
            preorder(root.left);
            preorder(root.right);
        }
    }

    void postorder(Node root) {
        if (root != null) {
            postorder(root.left);
            postorder(root.right);
            System.out.print(root.data + " ");
        }
    }

    void levelOrder() {
        Queue<Node> q = new LinkedList<>();
        q.add(root);
        while (!q.isEmpty()) {
            Node temp = q.poll();
            System.out.print(temp.data + " ");
            if (temp.left != null) q.add(temp.left);
            if (temp.right != null) q.add(temp.right);
        }
    }

    public static void main(String[] args) {
        BinaryTree bt = new BinaryTree();
        bt.insert(10);
        bt.insert(20);
        bt.insert(30);
        bt.insert(40);

        System.out.print("Inorder: ");
        bt.inorder(bt.root);

        System.out.print("\nPreorder: ");
        bt.preorder(bt.root);

        System.out.print("\nPostorder: ");
        bt.postorder(bt.root);

        System.out.print("\nLevel Order: ");
        bt.levelOrder();

        System.out.println("\nSearch 30: " + bt.search(bt.root, 30));
    }
}

File: trees/binary_tree.py
# File Name: binary_tree.py
from collections import deque

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self):
        self.root = None

    def insert(self, data):
        new_node = Node(data)
        if not self.root:
            self.root = new_node
            return

        q = deque([self.root])
        while q:
            temp = q.popleft()
            if not temp.left:
                temp.left = new_node
                return
            else:
                q.append(temp.left)

            if not temp.right:
                temp.right = new_node
                return
            else:
                q.append(temp.right)

    def search(self, root, key):
        if not root:
            return False
        if root.data == key:
            return True
        return self.search(root.left, key) or self.search(root.right, key)

    def inorder(self, root):
        if root:
            self.inorder(root.left)
            print(root.data, end=" ")
            self.inorder(root.right)

    def preorder(self, root):
        if root:
            print(root.data, end=" ")
            self.preorder(root.left)
            self.preorder(root.right)

    def postorder(self, root):
        if root:
            self.postorder(root.left)
            self.postorder(root.right)
            print(root.data, end=" ")

    def level_order(self):
        q = deque([self.root])
        while q:
            temp = q.popleft()
            print(temp.data, end=" ")
            if temp.left:
                q.append(temp.left)
            if temp.right:
                q.append(temp.right)

bt = BinaryTree()
bt.insert(10)
bt.insert(20)
bt.insert(30)
bt.insert(40)

print("Inorder:", end=" ")
bt.inorder(bt.root)

print("\nPreorder:", end=" ")
bt.preorder(bt.root)

print("\nPostorder:", end=" ")
bt.postorder(bt.root)

print("\nLevel Order:", end=" ")
bt.level_order()

print("\nSearch 30:", bt.search(bt.root, 30))
